Sources/Phoenix/Socket.swift

```diff
@@ -211,32 +213,73 @@ extension Socket: ConnectablePublisher {
     }
     
     public func disconnect() {
-        sync {
-            self.shouldReconnect = false
-            
-            self.cancelHeartbeatTimer()
+        // Calling `Channel.leave()` inside `sync` can cause a deadlock.
+        let channels: [Channel] = sync {
+            os_log(
+                "socket.disconnect: oldstate=%{public}@",
+                log: .phoenix,
+                type: .debug,
+                state.debugDescription
+            )
+
+            shouldReconnect = false
+            cancelHeartbeatTimer()
             
             switch state {
             case .closed, .closing:
-                // NOOP
-                return
+                break
             case .open(let ws), .connecting(let ws):
-                self.state = .closing(ws)
+                state = .closing(ws)
 
                 let subject = self.subject
                 notifySubjectQueue.async { subject.send(.closing) }
 
                 ws.close()
             }
+
+            return self.channels.compactMap { $0.value.channel }
+        }
+        channels.forEach { $0.leave() }
+    }
+
+    @discardableResult
```

Comment:
@myobie This is what I came up with for our be-sure-the-socket-is-closed-and-gone version of `disconnect()`.